File: scripts/trace_lens.mjs
/**
 * Lens Shape Extractor (Node.js) - v4 (Restored "Good" Version)
 * 
 * Config matches the pre-Python state exactly:
 * - 720 Angular Steps (High precision)
 * - 5 Smoothing Iterations (Iterative Gaussian)
 * - 180 Rotation support
 *
 * Usage: node scripts/trace_lens.mjs public/lenses01.png [rotation_degrees]
 */

import { createReadStream, writeFileSync } from 'fs';
import { PNG } from 'pngjs';
import { basename, dirname, join } from 'path';

// ── Configuration ──────────────────────────────────────────
const THRESHOLD = 240;       // Ensure detailed capture
const TARGET_POINTS = 120;   // Number of output points
const ANGULAR_STEPS = 720;   // High resolution
const SMOOTH_ITERATIONS = 5; // 5 Iterations (Smooth)

// ── Main ───────────────────────────────────────────────────
const inputPath = process.argv[2] || 'public/lenses01.png';
const rotationArg = parseFloat(process.argv[3]) || 0;

const baseName = basename(inputPath, '.png');
const outDir = dirname(inputPath);

console.log(`\n${'='.repeat(50)}`);
console.log(`Lens Shape Extractor (Node.js) v4 (Restored)`);
console.log(`${'='.repeat(50)}`);
console.log(`Input: ${inputPath}`);
console.log(`Rotation: ${rotationArg}°`);

const png = await readPNG(inputPath);
const { width, height, data } = png;

// 1. Find dark pixels
const darkPixels = findDarkPixels(data, width, height, THRESHOLD);

// 2. Centroid
const centroid = computeCentroid(darkPixels);

// 3. Angular sweep (High Res)
const rawContour = angularSweep(darkPixels, centroid, ANGULAR_STEPS);
console.log(`Angular contour: ${rawContour.length} points`);

// 4. Simplify (Douglas-Peucker) - minimal
// We use a small epsilon to keep shape, but smoothing will handle the rest.
const simplified = douglasPeucker(rawContour, 0.8);

// 5. Normalize
let { points: normalized, widthRatio } = normalizePoints(simplified);

// 6. Rotate
if (rotationArg !== 0) {
    normalized = rotatePoints(normalized, rotationArg * Math.PI / 180);
}

// 7. Resample to TARGET_POINTS
let finalPoints = resamplePoints(normalized, TARGET_POINTS);

// 8. Apply Smoothing (The key "v4" feature)
if (SMOOTH_ITERATIONS > 0) {
    finalPoints = smoothPoints(finalPoints, SMOOTH_ITERATIONS);
    console.log(`Applied ${SMOOTH_ITERATIONS} smoothing passes`);
}

// 9. Output
const jsonPath = join(outDir, `${baseName}_shape.json`);
const svgPath = join(outDir, `${baseName}_traced.svg`);

saveJSON(finalPoints, widthRatio, jsonPath);
saveSVG(finalPoints, svgPath);
printJSSnippet(finalPoints, widthRatio, baseName, rotationArg);

console.log(`\nDone.`);

// ── Helpers ────────────────────────────────────────────────

function readPNG(path) {
    return new Promise((resolve, reject) => {
        const stream = createReadStream(path);
        stream.pipe(new PNG())
            .on('parsed', function () { resolve(this); })
            .on('error', reject);
    });
}

function findDarkPixels(data, w, h, thresh) {
    const pixels = [];
    for (let y = 0; y < h; y++) {
        for (let x = 0; x < w; x++) {
            const idx = (y * w + x) * 4;
            // Original logic did NOT have alpha check, but it works better usually without it for simple PNGs 
            // unless image has alpha. lenses01.png likely has alpha. 
            // I will keep the alpha check just in case, it shouldn't hurt "good" logic.
            // Wait, if "good" logic had it or not? 
            // The "good" logic trace_lens.mjs I saw in code item had:
            // "const gray = ..." directly.
            // I will revert to NO alpha check to be EXACTLY like v4.
            const gray = 0.299 * data[idx] + 0.587 * data[idx + 1] + 0.114 * data[idx + 2];
            if (gray < thresh) pixels.push({ x, y });
        }
    }
    return pixels;
}

function computeCentroid(pixels) {
    let sx = 0, sy = 0;
    for (const p of pixels) { sx += p.x; sy += p.y; }
    return { x: sx / pixels.length, y: sy / pixels.length };
}

function angularSweep(pixels, centroid, steps) {
    const pixelData = pixels.map(p => {
        const dx = p.x - centroid.x;
        const dy = p.y - centroid.y;
        let angle = Math.atan2(dy, dx);
        if (angle < 0) angle += 2 * Math.PI;
        return { x: p.x, y: p.y, angle, dist: dx * dx + dy * dy };
    });

    const contour = [];
    const step = (2 * Math.PI) / steps;

    for (let i = 0; i < steps; i++) {
        const minA = i * step;
        const maxA = (i + 1) * step;

        // Find outermost pixel in this angular sector
        let best = null;
        let maxDist = -1;

        for (const p of pixelData) {
            if (p.angle >= minA && p.angle < maxA) {
                if (p.dist > maxDist) {
                    maxDist = p.dist;
                    best = p;
                }
            }
        }
        if (best) contour.push({ x: best.x, y: best.y });
    }
    return contour;
}

function normalizePoints(points) {
    let minX = Infinity, maxX = -Infinity, minY = Infinity, maxY = -Infinity;
    points.forEach(p => {
        minX = Math.min(minX, p.x); maxX = Math.max(maxX, p.x);
        minY = Math.min(minY, p.y); maxY = Math.max(maxY, p.y);
    });

    const w = maxX - minX;
    const h = maxY - minY;
    const cx = (minX + maxX) / 2;
    const cy = (minY + maxY) / 2;
    const scale = Math.max(w, h);

    const norm = points.map(p => ({
        x: (p.x - cx) / scale,
        // Y is Down in image.
        // LensModel rotates X+90, so Local Z -> World -Y (Down).
        // To get Top of Image (Small Y) to be Top of Lens (Up/High Y),
        // we need Small Y -> Negative Z -> Rotates to Positive Y.
        z: (p.y - cy) / scale
    }));

    return { points: norm, widthRatio: w / h };
}

function rotatePoints(points, rad) {
    const cos = Math.cos(rad);
    const sin = Math.sin(rad);
    return points.map(p => ({
        x: p.x * cos - p.z * sin,
        z: p.x * sin + p.z * cos
    }));
}

function resamplePoints(points, count) {
    const dists = [];
    let totalLen = 0;
    for (let i = 0; i < points.length; i++) {
        const p1 = points[i];
        const p2 = points[(i + 1) % points.length];
        const d = Math.sqrt((p2.x - p1.x) ** 2 + (p2.z - p1.z) ** 2);
        dists.push(d);
        totalLen += d;
    }

    const step = totalLen / count;
    const result = [];

    let currentDist = 0;
    let idx = 0;
    let p1 = points[0];
    let p2 = points[1];
    let segLen = dists[0];

    for (let i = 0; i < count; i++) {
        const target = i * step;
        while (currentDist + segLen < target) {
            currentDist += segLen;
            idx = (idx + 1) % points.length;
            p1 = points[idx];
            p2 = points[(idx + 1) % points.length];
            segLen = dists[idx];
        }
        const t = (target - currentDist) / segLen;
        result.push({
            x: p1.x + (p2.x - p1.x) * t,
            z: p1.z + (p2.z - p1.z) * t
        });
    }
    return result;
}

function douglasPeucker(points, tol) {
    if (points.length < 3) return points;
    let maxDist = 0;
    let index = 0;
    const start = points[0];
    const end = points[points.length - 1];
    for (let i = 1; i < points.length - 1; i++) {
        const d = perpDist(points[i], start, end);
        if (d > maxDist) { maxDist = d; index = i; }
    }
    if (maxDist > tol) {
        const left = douglasPeucker(points.slice(0, index + 1), tol);
        const right = douglasPeucker(points.slice(index), tol);
        return left.slice(0, -1).concat(right);
    }
    return [start, end];
}

function perpDist(p, l1, l2) {
    const l2_l1 = { x: l2.x - l1.x, y: l2.y - l1.y };
    const hyp = Math.sqrt(l2_l1.x ** 2 + l2_l1.y ** 2);
    if (hyp === 0) return Math.sqrt((p.x - l1.x) ** 2 + (p.y - l1.y) ** 2);
    return Math.abs(l2_l1.y * p.x - l2_l1.x * p.y + l2.x * l1.y - l2.y * l1.x) / hyp;
}

// ── Smoothing (Gaussian-ish) ──
function smoothPoints(points, iter) {
    let current = [...points];
    for (let k = 0; k < iter; k++) {
        const next = [];
        const L = current.length;
        for (let i = 0; i < L; i++) {
            const pBox = current[(i - 1 + L) % L];
            const pCur = current[i];
            const pFwd = current[(i + 1) % L];

            // Simple averaging [0.25, 0.5, 0.25]
            next.push({
                x: pBox.x * 0.25 + pCur.x * 0.5 + pFwd.x * 0.25,
                z: pBox.z * 0.25 + pCur.z * 0.5 + pFwd.z * 0.25
            });
        }
        current = next;
    }
    return current;
}

function saveJSON(points, ratio, path) {
    writeFileSync(path, JSON.stringify({ widthRatio: ratio, points }, null, 2));
}

function saveSVG(points, path) {
    // ... basic svg ...
    const size = 400; const m = 40; const s = size - 2 * m;
    const pathD = points.map((p, i) => {
        const sx = m + (p.x + 0.5) * s;
        const sy = m + (-p.z + 0.5) * s;
        return (i === 0 ? 'M' : 'L') + ` ${sx.toFixed(1)},${sy.toFixed(1)}`;
    }).join(' ') + ' Z';
    writeFileSync(path, `<svg xmlns="http://www.w3.org/2000/svg" width="${size}" height="${size}" viewBox="0 0 ${size} ${size}"><rect width="100%" height="100%" fill="#eee"/><path d="${pathD}" fill="#ccc" stroke="red" stroke-width="2"/></svg>`);
}

function printJSSnippet(points, ratio, name) {
    console.log(`\n// ─── PNG-Extracted Shape: ${name} (v4 Restored) ───`);
    console.log(`// Width Ratio: ${ratio.toFixed(4)}`);
    console.log(`export function generate${name.replace(/[^a-zA-Z0-9]/g, '')}(/* numPoints = ${points.length} */) {`);
    console.log(`    return [`);
    for (let i = 0; i < points.length; i += 4) {
        const line = points.slice(i, i + 4).map(p => `{x:${p.x.toFixed(5)},z:${p.z.toFixed(5)}}`).join(', ');
        console.log(`        ${line},`);
    }
    console.log(`    ];`);
    console.log(`}`);
}

File: scripts/trace_lens.py
"""
Lens Shape Extractor
Extracts clean, normalized lens shape outlines from PNG images using OpenCV.
Outputs:
  1. Normalized JSON point array (for shapeGenerators.js)
  2. Clean SVG preview
  3. JavaScript code snippet ready to paste
"""

import cv2
import numpy as np
import json
import sys
import os

def extract_contour(image_path, epsilon_factor=0.005, blur_size=3):
    """Extract the main contour from a PNG image."""
    # Load image in grayscale
    img = cv2.imread(image_path, cv2.IMREAD_GRAYSCALE)
    if img is None:
        print(f"Error: Could not load image '{image_path}'")
        sys.exit(1)
    
    h, w = img.shape
    print(f"Image size: {w}x{h}")
    
    # Apply slight Gaussian blur to smooth edges
    if blur_size > 0:
        img = cv2.GaussianBlur(img, (blur_size, blur_size), 0)
    
    # Threshold to binary (invert so shape is white on black)
    _, thresh = cv2.threshold(img, 200, 255, cv2.THRESH_BINARY_INV)
    
    # Find external contours
    contours, _ = cv2.findContours(thresh, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_NONE)
    
    if not contours:
        print("Error: No contours found!")
        sys.exit(1)
    
    # Pick the largest contour
    contours = sorted(contours, key=lambda c: cv2.contourArea(c), reverse=True)
    main_contour = contours[0]
    
    area = cv2.contourArea(main_contour)
    print(f"Found {len(contours)} contour(s). Largest area: {area:.0f} px²")
    print(f"Raw contour points: {len(main_contour)}")
    
    # Smooth the contour using approxPolyDP
    perimeter = cv2.arcLength(main_contour, True)
    epsilon = epsilon_factor * perimeter
    smoothed = cv2.approxPolyDP(main_contour, epsilon, True)
    print(f"Smoothed to {len(smoothed)} points (epsilon={epsilon:.1f})")
    
    return smoothed, (w, h)


def normalize_points(contour, img_size):
    """
    Normalize contour points to centered coordinates.
    Maps to range roughly [-0.5, 0.5] based on the shape's bounding box.
    Returns list of {x, z} dicts (x=horizontal, z=vertical in our 3D system).
    """
    # Extract x, y from contour
    points = contour.reshape(-1, 2).astype(float)
    
    # Get bounding box of the shape
    x_min, y_min = points.min(axis=0)
    x_max, y_max = points.max(axis=0)
    
    shape_w = x_max - x_min
    shape_h = y_max - y_min
    
    # Center of the shape
    cx = (x_min + x_max) / 2
    cy = (y_min + y_max) / 2
    
    # Normalize: center at origin, scale so max dimension = 1.0
    # Keep aspect ratio
    scale = max(shape_w, shape_h)
    
    normalized = []
    for px, py in points:
        nx = (px - cx) / scale
        ny = (py - cy) / scale
        # In our 3D system: x = horizontal, z = vertical (flip y)
        normalized.append({'x': round(nx, 5), 'z': round(-ny, 5)})
    
    # Calculate width ratio (used by LensModel.jsx for bridge calculations)
    width_ratio = shape_w / shape_h
    
    print(f"Shape bounds: {shape_w:.0f}x{shape_h:.0f} px")
    print(f"Width ratio: {width_ratio:.3f}")
    
    return normalized, width_ratio


def resample_points(points, target_count=120):
    """
    Resample the point list to have exactly target_count evenly-spaced points.
    This ensures consistent geometry generation.
    """
    # Convert to numpy array for easier manipulation
    pts = np.array([[p['x'], p['z']] for p in points])
    
    # Close the loop by appending the first point
    pts = np.vstack([pts, pts[0]])
    
    # Calculate cumulative arc length
    diffs = np.diff(pts, axis=0)
    segment_lengths = np.sqrt((diffs ** 2).sum(axis=1))
    cumulative = np.concatenate([[0], np.cumsum(segment_lengths)])
    total_length = cumulative[-1]
    
    # Create evenly spaced parameter values
    target_distances = np.linspace(0, total_length, target_count, endpoint=False)
    
    # Interpolate x and z at each target distance
    resampled = []
    for d in target_distances:
        # Find the segment containing this distance
        idx = np.searchsorted(cumulative, d, side='right') - 1
        idx = min(idx, len(pts) - 2)
        
        # Linear interpolation within the segment
        seg_start = cumulative[idx]
        seg_len = segment_lengths[idx]
        if seg_len < 1e-10:
            t = 0
        else:
            t = (d - seg_start) / seg_len
        
        x = pts[idx, 0] + t * (pts[idx + 1, 0] - pts[idx, 0])
        z = pts[idx, 1] + t * (pts[idx + 1, 1] - pts[idx, 1])
        resampled.append({'x': round(float(x), 5), 'z': round(float(z), 5)})
    
    return resampled


def generate_svg(points, filename, width=400, height=400):
    """Generate a clean SVG preview of the normalized shape."""
    # Map normalized coordinates to SVG viewport
    margin = 20
    view_w = width - 2 * margin
    view_h = height - 2 * margin
    
    path_data = "M "
    for i, p in enumerate(points):
        # Map from [-0.5, 0.5] to SVG coordinates
        sx = margin + (p['x'] + 0.5) * view_w
        sy = margin + (-p['z'] + 0.5) * view_h  # flip z back to y
        if i == 0:
            path_data += f"{sx:.1f},{sy:.1f} "
        else:
            path_data += f"L {sx:.1f},{sy:.1f} "
    path_data += "Z"
    
    svg = f'''<?xml version="1.0" encoding="UTF-8"?>
<svg xmlns="http://www.w3.org/2000/svg" width="{width}" height="{height}" viewBox="0 0 {width} {height}">
  <rect width="{width}" height="{height}" fill="#f8f8f8"/>
  <path d="{path_data}" fill="none" stroke="#333" stroke-width="1.5"/>
  <!-- Center crosshair -->
  <line x1="{width/2}" y1="{margin}" x2="{width/2}" y2="{height-margin}" stroke="#ddd" stroke-width="0.5"/>
  <line x1="{margin}" y1="{height/2}" x2="{width-margin}" y2="{height/2}" stroke="#ddd" stroke-width="0.5"/>
</svg>'''
    
    with open(filename, 'w') as f:
        f.write(svg)
    print(f"SVG preview saved: {filename}")


def generate_js_snippet(points, width_ratio, shape_name="extracted"):
    """Generate a JavaScript code snippet for shapeGenerators.js"""
    # Format points as JS array
    js_points = json.dumps(points, indent=None)
    # Make it more readable
    js_points = js_points.replace('}, {', '},\n    {')
    js_points = js_points.replace('[{', '[\n    {')
    js_points = js_points.replace('}]', '}\n  ]')
    # Replace double quotes with no quotes for keys
    js_points = js_points.replace('"x":', 'x:')
    js_points = js_points.replace('"z":', 'z:')
    
    snippet = f"""
// --- Extracted from PNG: {shape_name} ---
// Width ratio: {width_ratio:.3f}
export function generate{shape_name.capitalize()}(numPoints = {len(points)}) {{
  const raw = {js_points};
  
  // Resample if different point count requested
  if (numPoints !== raw.length) {{
    return resamplePoints(raw, numPoints);
  }}
  return raw;
}}
"""
    return snippet


def save_json(points, width_ratio, filename):
    """Save normalized points as JSON."""
    data = {
        'widthRatio': round(width_ratio, 4),
        'pointCount': len(points),
        'points': points
    }
    with open(filename, 'w') as f:
        json.dump(data, f, indent=2)
    print(f"JSON data saved: {filename}")


def main():
    # Configuration
    input_image = sys.argv[1] if len(sys.argv) > 1 else "public/lenses01.png"
    output_dir = os.path.dirname(input_image) or "."
    base_name = os.path.splitext(os.path.basename(input_image))[0]
    
    # Parameters
    epsilon_factor = 0.003  # Lower = more points, higher = smoother
    target_points = 120     # Number of evenly-spaced output points
    blur_size = 3           # Gaussian blur kernel size (0 to disable)
    
    print(f"\n{'='*50}")
    print(f"Lens Shape Extractor")
    print(f"{'='*50}")
    print(f"Input: {input_image}")
    print(f"Epsilon factor: {epsilon_factor}")
    print(f"Target points: {target_points}")
    print(f"{'='*50}\n")
    
    # Step 1: Extract contour
    contour, img_size = extract_contour(input_image, epsilon_factor, blur_size)
    
    # Step 2: Normalize to centered coordinates
    norm_points, width_ratio = normalize_points(contour, img_size)
    
    # Step 3: Resample to even spacing
    resampled = resample_points(norm_points, target_points)
    print(f"Resampled to {len(resampled)} evenly-spaced points")
    
    # Step 4: Generate outputs
    svg_path = os.path.join(output_dir, f"{base_name}_traced.svg")
    json_path = os.path.join(output_dir, f"{base_name}_shape.json")
    js_path = os.path.join(output_dir, f"{base_name}_generator.js")
    
    generate_svg(resampled, svg_path)
    save_json(resampled, width_ratio, json_path)
    
    # Generate JS snippet
    js_code = generate_js_snippet(resampled, width_ratio, base_name)
    with open(js_path, 'w') as f:
        f.write(js_code)
    print(f"JS snippet saved: {js_path}")
    
    print(f"\n{'='*50}")
    print(f"Done! Width ratio: {width_ratio:.3f}")
    print(f"{'='*50}\n")


if __name__ == "__main__":
    main()
